README.md
# record_amr

record and play amr file.

## Getting Started

Android:

```
<uses-permission android:name="android.permission.MODIFY_AUDIO_SETTINGS" />
<uses-permission android:name="android.permission.RECORD_AUDIO" />
```

iOS:



```
Privacy - Microphone Usage Description
```
change `GeneratedPluginRegistrant.m` to `GeneratedPluginRegistrant.mm`

CHANGELOG.md
## 0.0.1

* android and iOS support record amr files.
* android and iOS support play amr files.


test/record_amr_test.dart
import 'package:flutter/services.dart';
import 'package:flutter_test/flutter_test.dart';
import 'package:record_amr/record_amr.dart';

void main() {
  const MethodChannel channel = MethodChannel('record_amr');

  TestWidgetsFlutterBinding.ensureInitialized();

  setUp(() {
    channel.setMockMethodCallHandler((MethodCall methodCall) async {
      return '42';
    });
  });

  tearDown(() {
    channel.setMockMethodCallHandler(null);
  });

  test('getPlatformVersion', () async {
    expect(await RecordAmr.platformVersion, '42');
  });
}


example/test/widget_test.dart
// This is a basic Flutter widget test.
//
// To perform an interaction with a widget in your test, use the WidgetTester
// utility that Flutter provides. For example, you can send tap and scroll
// gestures. You can also use WidgetTester to find child widgets in the widget
// tree, read text, and verify that the values of widget properties are correct.

import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';

import 'package:record_amr_example/main.dart';

void main() {
  testWidgets('Verify Platform version', (WidgetTester tester) async {
    // Build our app and trigger a frame.
    await tester.pumpWidget(MyApp());

    // Verify that platform version is retrieved.
    expect(
      find.byWidgetPredicate(
        (Widget widget) => widget is Text &&
                           widget.data.startsWith('Running on:'),
      ),
      findsOneWidget,
    );
  });
}


lib/record_amr.dart
import 'dart:async';

import 'package:flutter/services.dart';

/// `volume` 0 ~ 1;
typedef VolumeCallBack = Function(double volume);
typedef StopRecordCallBack = void Function(String path, int duration);
typedef StopPlayCallBack = void Function(String path);

class RecordAmr {
  static const MethodChannel _channel = const MethodChannel('record_amr');

  static RecordAmr _recordAmr;

  // ignore: unused_field
  VolumeCallBack _callBack;
  StopPlayCallBack _stopCallBack;

  static RecordAmr get _private => _recordAmr = _recordAmr ?? RecordAmr._();

  RecordAmr._() {
    _channel.setMethodCallHandler((call) {
      if (call.method == 'volume') {
        double volume = call.arguments.toDouble();
        if (_private._callBack != null) {
          _private._callBack(volume);
        }
      }

      if (call.method == 'stopPlaying') {
        String path = call.arguments["path"] as String;
        if (_private._stopCallBack != null) {
          _private._stopCallBack(path);
          _private._stopCallBack = null;
        }
      }
      return null;
    });
  }

  bool recoreding = false;
  VolumeCallBack callback;

  /// start record
  /// [path] record file path.
  /// [callBack] volume callback: 0 ~ 1.
  static Future<bool> startVoiceRecord([
    VolumeCallBack volumeCallBack,
  ]) async {
    if (_private.recoreding) {
      return false;
    }
    if (volumeCallBack != null) {
      _private._callBack = volumeCallBack;
    }

    _private.recoreding =
        await _channel.invokeMethod('startVoiceRecord') as bool;

    return _private.recoreding;
  }

  /// stop record
  static Future<bool> stopVoiceRecord(
    StopRecordCallBack callBack,
  ) async {
    Map result = await _channel.invokeMethod('stopVoiceRecord');
    _private._callBack = null;
    _private.recoreding = false;
    String error = result['error'];
    String path = result['path'];
    int duration = result['duration'] as int;
    callBack(path, duration);
    return error == null ? true : false;
  }

  /// cancel record
  static Future<Null> cancelVoiceRecord() async {
    await _channel.invokeMethod('cancelVoiceRecord');
    _private._callBack = null;
    _private.recoreding = false;
  }

  /// play amr file
  static Future<bool> play(
    String path, [
    StopPlayCallBack endCallback,
  ]) async {
    _private._stopCallBack = endCallback;
    bool isPlay = await _channel.invokeMethod('play', {"path": path}) as bool;
    return isPlay;
  }

  /// stop playing amr file
  static Future<Null> stop() async {
    _private._stopCallBack = null;
    await _channel.invokeMethod('stopPlaying') as bool;
    return null;
  }
}


example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:record_amr/record_amr.dart';
import 'package:permission_handler/permission_handler.dart';

void main() {
  runApp(MyApp());
}

class MyApp extends StatefulWidget {
  @override
  _MyAppState createState() => _MyAppState();
}

class _MyAppState extends State<MyApp> {
  double _volume = 0;
  String _path;

  @override
  void initState() {
    _requestPermiss();
    super.initState();
  }

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      home: Scaffold(
        appBar: AppBar(
          title: const Text('Plugin example app'),
        ),
        body: Center(
          child: Column(
            crossAxisAlignment: CrossAxisAlignment.center,
            children: [
              Container(
                width: _volume * 100,
                height: 30,
                color: Colors.blue[300],
              ),
              Expanded(
                child: GestureDetector(
                  onTapDown: (TapDownDetails details) => _startRecord(),
                  onTapUp: (TapUpDetails details) => _stopRecord(),
                  onTapCancel: () => _cancelRecord(),
                  child: Text(
                    'Record',
                  ),
                ),
              ),
              Row(
                mainAxisAlignment: MainAxisAlignment.center,
                children: [
                  IconButton(
                    icon: Icon(Icons.play_arrow),
                    onPressed: () => _play(),
                  ),
                  IconButton(
                    icon: Icon(Icons.stop),
                    onPressed: () => _stop(),
                  ),
                ],
              )
            ],
          ),
        ),
      ),
    );
  }

  _startRecord() async {
    bool success = await RecordAmr.startVoiceRecord((volume) {
      print('volume ---- $volume');
      setState(() {
        _volume = volume;
      });
    });
    print('start record ---- $success');
  }

  _stopRecord() async {
    bool success = await RecordAmr.stopVoiceRecord((path, duration) {
      _path = path;
      print('path --- $path, duration ---- $duration');
    });
    setState(() {
      _volume = 0;
    });
    print('stop record ---- $success');
  }

  _cancelRecord() async {
    await RecordAmr.cancelVoiceRecord();
    setState(() {
      _volume = 0;
    });
    print('取消录制');
  }

  _play() async {
    await RecordAmr.play(_path, (path) {
      print('play end');
    });
  }

  _stop() async {
    await RecordAmr.stop();
  }

  _requestPermiss() async {
    await Permission.microphone.request();
  }
}


ios/Classes/RecordAmrPlugin.m
#import "RecordAmrPlugin.h"
#import <AVFoundation/AVFoundation.h>
#import <EMVoiceConvert/EMVoiceConvert.h>

@interface RecordAmrPlugin ()
@property (nonatomic, strong) NSString *path;
@property (nonatomic, strong) AVAudioRecorder *recorder;
@end

@implementation RecordAmrPlugin


+ (void)registerWithRegistrar:(NSObject<FlutterPluginRegistrar>*)registrar {
    FlutterMethodChannel* channel = [FlutterMethodChannel
                                     methodChannelWithName:@"record_amr"
                                     binaryMessenger:[registrar messenger]];
    RecordAmrPlugin* instance = [[RecordAmrPlugin alloc] init];
    [registrar addMethodCallDelegate:instance channel:channel];
}

- (void)handleMethodCall:(FlutterMethodCall*)call
                  result:(FlutterResult)result {
    
    NSLog(@"call.method --- %@",call.method);
    
    if ([@"startVoiceRecord" isEqualToString:call.method])
    {
        [self startVoiceRecord:call.arguments result:result];
    } else if ([@"stopVoiceRecord" isEqualToString:call.method])
    {
        [self stopVoiceRecord:call.arguments result:result];
    } else if ([@"cancelVoiceRecord" isEqualToString:call.method])
    {
        [self cancelVoiceRecord:call.arguments result:result];
    } else {
        result(FlutterMethodNotImplemented);
    }
}

- (void)startVoiceRecord:(NSDictionary *)callInfo result:(FlutterResult)result {
    NSString *recordPath = [self.path stringByAppendingFormat:@"/%.0f", [[NSDate date] timeIntervalSince1970] * 1000];
    NSError *error = nil;
    [[AVAudioSession sharedInstance] setCategory:AVAudioSessionCategoryPlayAndRecord withOptions:AVAudioSessionCategoryOptionDuckOthers error:&error];
    if (!error){
        [[AVAudioSession sharedInstance] setActive:YES withOptions:AVAudioSessionSetActiveOptionNotifyOthersOnDeactivation error:&error];
    }
    
    if (error) {
        error = [NSError errorWithDomain:@"AVAudioSession SetCategory失败" code:-1 userInfo:nil];
        break;
    }
    NSLog(@"start record");
    result(@(YES));
}

- (void)stopVoiceRecord:(NSDictionary *)callInfo result:(FlutterResult)result {
    NSLog(@"stop record");
    result(@"我是路径");
}

- (void)cancelVoiceRecord:(NSDictionary *)callInfo result:(FlutterResult)result {
    result(@(YES));
}


- (NSString *)path
{
    if (!_path) {
        _path =  [NSSearchPathForDirectoriesInDomains(NSCachesDirectory, NSUserDomainMask, YES) firstObject];
        _path = [_path stringByAppendingPathComponent:@"EMRecord"];
        if (![[NSFileManager defaultManager] fileExistsAtPath:_path]) {
            [[NSFileManager defaultManager] createDirectoryAtPath:_path withIntermediateDirectories:YES attributes:nil error:nil];
        }
    }
    return _path;
}
@end


android/src/main/java/com/easemob/djp/record_amr/RecordAmrPlugin.java
package com.easemob.djp.record_amr;

import android.content.Context;
import android.media.MediaPlayer;
import android.media.MediaRecorder;

import android.net.Uri;
import android.os.Handler;
import android.os.Looper;

import org.json.JSONException;
import org.json.JSONObject;

import java.io.File;

import java.io.IOException;
import java.util.HashMap;
import java.util.Map;

import androidx.annotation.NonNull;

import io.flutter.embedding.engine.plugins.FlutterPlugin;
import io.flutter.plugin.common.MethodCall;
import io.flutter.plugin.common.MethodChannel;
import io.flutter.plugin.common.MethodChannel.MethodCallHandler;
import io.flutter.plugin.common.MethodChannel.Result;


/** RecordAmrPlugin */
public class RecordAmrPlugin implements FlutterPlugin, MethodCallHandler, MediaPlayer.OnCompletionListener {
  private static final String TAG = "RecordAmr";

  private MethodChannel channel;
  //文件路径
  private String filePath;
  //文件夹路径
  private String FolderPath = "/EMRecord/";

  private MediaRecorder mMediaRecorder;

  private MediaPlayer mediaPlayer;

  private Context context;


  private Handler mainHandler = new Handler(Looper.getMainLooper());

  @Override
  public void onAttachedToEngine(@NonNull FlutterPluginBinding flutterPluginBinding) {
    channel = new MethodChannel(flutterPluginBinding.getBinaryMessenger(), "record_amr");
    channel.setMethodCallHandler(this);
    context = flutterPluginBinding.getApplicationContext();
  }

  @Override
  public void onMethodCall(@NonNull MethodCall call, @NonNull Result result) {
    if (call.method.equals("startVoiceRecord")) {
      this.startRecord(result);
    }
    else if (call.method.equals("stopVoiceRecord")) {
      this.stopRecord(result);
    }
    else if (call.method.equals("cancelVoiceRecord")) {
      this.cancelRecord(result);
    }
    else if (call.method.equals("play")) {
      HashMap map = (HashMap) call.arguments;
      String path = (String) map.get("path");
      this.play(path, result);
    }
    else if (call.method.equals("stopPlaying")) {
      this.stop(result);
    }
   else{
      result.notImplemented();
    }
  }

  @Override
  public void onDetachedFromEngine(@NonNull FlutterPluginBinding binding) {
    channel.setMethodCallHandler(null);
  }

  private long startTime;

  // 开始录音
  private void startRecord(Result result) {
    if (mMediaRecorder == null)
      mMediaRecorder = new MediaRecorder();
    try {
      startTime = System.currentTimeMillis();
      mMediaRecorder.setAudioSource(MediaRecorder.AudioSource.MIC);// 设置麦克风
      mMediaRecorder.setOutputFormat(MediaRecorder.OutputFormat.DEFAULT);
      mMediaRecorder.setAudioEncoder(MediaRecorder.AudioEncoder.AMR_NB);
      filePath = context.getFilesDir().getPath() + "/" + startTime + ".amr" ;
      mMediaRecorder.setOutputFile(filePath);
      mMediaRecorder.prepare();
      mMediaRecorder.start();
      updateMicStatus();
      result.success(Boolean.valueOf(true));
    } catch (Exception e) {
      mMediaRecorder.reset();
      mMediaRecorder.release();
      mMediaRecorder = null;
      filePath = "";
      result.success(Boolean.valueOf(false));
    }
  }

  // 停止录音
  private void stopRecord(Result result) {

    Map<String, Object> map = new HashMap<>();
    map.put("duration", 0L);
    map.put("path", "");
    do {
      try {
        if (mMediaRecorder == null) {
          break;
        }else {
          mMediaRecorder.stop();
          map.put("duration", (System.currentTimeMillis() - startTime) / 1000);
          map.put("path", filePath);
          mMediaRecorder.reset();
          mMediaRecorder.release();
          filePath = "";
          break;
        }
      }catch (RuntimeException e){
        File file = new File(filePath);
        if (file.exists())
        {
          file.delete();
        }

      }finally {
        mMediaRecorder = null;
        filePath = "";
      }
    }while (false);

    result.success(map);
  }

  private void cancelRecord(Result result) {
    mMediaRecorder.stop();
    mMediaRecorder.reset();
    mMediaRecorder.release();
    mMediaRecorder = null;
    filePath = "";
    result.success(Boolean.valueOf(true));
  }

  private void play(String path, Result result) {

    if (mediaPlayer != null && filePath.equals(path)) {
      result.success(Boolean.TRUE);
      return;
    } else {
      if (filePath != null) {
        stop(null);
      }
    }
    this.stop(null);

    filePath = path;
    mediaPlayer = new  MediaPlayer();
    try {
      mediaPlayer.setDataSource(path);
      mediaPlayer.setOnCompletionListener(this);
      mediaPlayer.prepare();
      mediaPlayer.start();
      result.success(Boolean.TRUE);
    } catch (IOException e) {
      result.success(Boolean.FALSE);
      e.printStackTrace();
    }
  }

  private void stop(Result result) {
    if (mediaPlayer != null) {
      if (mediaPlayer.isPlaying() == true) {
          mediaPlayer.stop();
      }
      mediaPlayer.reset();
      mediaPlayer.release();
      mediaPlayer = null;
    }
    if (result != null) {
      HashMap map = new HashMap();
      map.put("error", Boolean.FALSE);
      map.put("path" , filePath != null ? filePath : "");
      channel.invokeMethod("stopPlaying", map);
    }
    filePath = null;
  }

  private double BASE = 3000;
  private int SPACE = 300;// 间隔取样时间

  // 获取麦克风音量大小
  private void updateMicStatus() {
    if (mMediaRecorder != null) {
      double ratio = (double)mMediaRecorder.getMaxAmplitude() / BASE;
      double db = 0;// 分贝
      if (ratio > 1) ratio = 1;
      db = ratio;
      final double finalDb = db;
      mainHandler.post(new Runnable() {
        @Override
        public void run() {
          channel.invokeMethod("volume", finalDb);
        }
      });
      mHandler.postDelayed(mUpdateMicStatusTimer, SPACE);
    }
  }

  private final Handler mHandler = new Handler();
  private Runnable mUpdateMicStatusTimer = new Runnable() {
    public void run() {
      updateMicStatus();
    }
  };

  @Override
  public void onCompletion(MediaPlayer mediaPlayer) {
    mediaPlayer.stop();
    mediaPlayer.reset();
    mediaPlayer.release();
    this.mediaPlayer = null;
    HashMap map = new HashMap();
    map.put("error", Boolean.FALSE);
    map.put("path" , filePath != null ? filePath : "");
    channel.invokeMethod("stopPlaying", map);
    filePath = null;
  }
}


